rspace/src/main/scala/coop/rchain/rspace/trace/Event.scala

```diff
@@ -38,6 +38,139 @@ object Event {
       }(Codec[Consume])
 
   implicit def codecLog: Codec[Seq[Event]] = codecSeq[Event](codecEvent)
+
+  /**
+    * `Scope` is widely used in descriptions in this file. To better understand `scope` here. I'd like to present
+    * some examples to demonstrate that.
+    *
+    * B2  Rho("for (_ <<- @1) { 0 }")
+    * |
+    * B1  Rho("@1!(0)")
+    *
+    * I'd like to use `Volatile` event here to demonstrate the `scope` here.
+    *
+    * Within the scope which Scope(B2, B3)
+    *
+    * B2  Rho("for (_ <<- @1) { 0 }") ----+
+    * |                                   |---Scope(B2, B1)
+    * B1  Rho("@1!(0)")               ----+
+    *
+    * If you look at within the scope of Scope(B2, B1) like above, all the comm events happened in B1 and B2 are Volatile.
+    *
+    *
+    * B2  Rho("for (_ <<- @1) { 0 }") ---- Scope(B2)
+    * |
+    * B1  Rho("@1!(0)")
+    *
+    * But if you look at within only the scope os Scope(B2) like above, the comm and consume events in the B2
+    * are not Volatile.
+    *
+    * `Volatile` means that the produce event and the consume event match within the scope.
+    * This produce and consume events within this comm event would not leave any thing left in the
+    * tuplespace within the scope.
+    */
+  def isVolatile(comm: COMM, consumes: Set[Consume], produces: Set[Produce]) =
+    !comm.consume.persistent && comm.peeks.isEmpty && consumes.contains(comm.consume) && comm.produces
+      .forall(
+        produce => !produce.persistent && produces.contains(produce)
+      )
+
+  /**
+    * @return EventGroup contains all the events which is not volatile in the scope
+    */
+  def extractRSpaceEventGroup(events: List[Event]): EventGroup = {
+    val rspaceEvents     = events.toSet
+    val allProduceEvents = rspaceEvents.collect { case p: Produce => p }
+    val allConsumeEvents = rspaceEvents.collect { case c: Consume => c }
+    val allCommEvents    = rspaceEvents.collect { case c: COMM => c }
```

Comment:
For better performance this can be simplified to fold only once over rspaceEvents and collect results in 3 groups.